keps/sig-node/4818-allow-zero-value-for-sleep-action-of-prestop-hook/README.md

```diff
@@ -0,0 +1,580 @@
+- Allow users to set a zero value for the sleep action in PreStop hooks to do a no-op.
+
+### Non-Goals
+
+- This KEP does not support adding negative values for the sleep duration.
+- This KEP does not aim to provide a way to pause or delay pod termination indefinitely.
+
+## Proposal
+
+Introduce a `PodLifecycleSleepActionAllowZero` feature gate which is disabled by default. When the feature gate is enabled, the `validateSleepAction` method would allow values greater than or equal to zero as a valid sleep duration.
+
+Since this update to the validation allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+
+- When the feature gate is disabled:
+  - (a) New resources will no longer allow setting zero as the sleep duration second for the PreStop hook. (no change to current validation)
+  - (b) Existing resources cannot be updated to have a sleep duration of zero seconds
+  - (c) Existing resources with a PreStop sleep duration set to zero will continue to run and use a sleep duration of zero seconds. These can be updated and the zero sleep duration would continue to work.
+- When the feature gate is enabled:
+  - (c) New resources allow zero as a valid sleep duration.
+  - (d) Updates to existing resources will allow zero as a valid sleep duration.
+
+The proposed change adds another layer to the `validateSleepAction` function to allow zero as a valid sleep duration setting like shown:
+
+```diff
+-func validateSleepAction(sleep *core.SleepAction, gracePeriod *int64, fldPath *field.Path) field.ErrorList {
++func validateSleepAction(sleep *core.SleepAction, gracePeriod *int64, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+	allErrors := field.ErrorList{}
+	// We allow gracePeriod to be nil here because the pod in which this SleepAction
+	// is defined might have an invalid grace period defined, and we don't want to
+	// flag another error here when the real problem will already be flagged.
+-	if gracePeriod != nil && sleep.Seconds <= 0 || sleep.Seconds > *gracePeriod {
+-		invalidStr := fmt.Sprintf("must be greater than 0 and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
+-		allErrors = append(allErrors, field.Invalid(fldPath, sleep.Seconds, invalidStr))
++	if opts.AllowPodLifecycleSleepActionZeroValue {
++		if gracePeriod != nil && sleep.Seconds < 0 || sleep.Seconds > *gracePeriod {
++			invalidStr := fmt.Sprintf("must be non-negative and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
++			allErrors = append(allErrors, field.Invalid(fldPath, sleep.Seconds, invalidStr))
++		}
++	} else {
++		if gracePeriod != nil && sleep.Seconds <= 0 || sleep.Seconds > *gracePeriod {
++			invalidStr := fmt.Sprintf("must be greater than 0 and less than terminationGracePeriodSeconds (%d)", *gracePeriod)
```

Comment:
Should we include a suggestion to include the new FG that can allow zero in the error message?